pkg/sql/lease.go

```diff
@@ -573,20 +584,57 @@ func (t *tableState) acquire(
 ) (*tableVersionState, error) {
 	t.mu.Lock()
 	defer t.mu.Unlock()
-	// Wait for any existing lease acquisition.
-	t.acquireWait()
 
-	// Acquire a lease if no lease exists or if the latest lease is
-	// about to expire.
+	if t.mu.acquiring != nil {
+		// Block and wait for the new lease to be acquired on another routine.
+		t.acquireWait()
+	}
+
+	// Ensure a lease is acquired. Even though we finished waiting, another
+	// routine which finished acquiring a lease could have released it by now.
 	if s := t.mu.active.findNewest(); s == nil || s.hasExpired(timestamp) {
 		if err := t.acquireNodeLease(ctx, m, hlc.Timestamp{}); err != nil {
 			return nil, err
 		}
 	}
 
+	if t.mu.timer == nil {
+		t.startRefreshingRoutine(m)
+	}
+
 	return t.findForTimestamp(ctx, timestamp, m)
 }
 
+// startRefreshingRoutine begins the goroutine responsible for refreshing a
+// table lease.
+//
+// t.mu needs to be locked.
+func (t *tableState) startRefreshingRoutine(m *LeaseManager) {
+	ctx := context.TODO()
+	t.mu.timer = time.NewTimer(LeaseRenewalTimeout)
+	if !t.mu.timer.Stop() {
+		<-t.mu.timer.C
+	}
+	t.stopper.RunWorker(ctx, func(ctx context.Context) {
+		for {
+			select {
+			case <-t.mu.timer.C:
+				t.mu.Lock()
+				// Get a new lease only if the currently newest one will expire soon.
+				// This lease refresh will be skipped if another acquision happens and
+				// the timer fires before it that routine resets the timer. Then we
+				// don't want to refresh the lease immediately after.
+				if s := t.mu.active.findNewest(); s == nil || s.willExpire(hlc.Timestamp{WallTime: hlc.UnixNano()}) {
```

Comment:
We really don't need to run these checks because it's just not possible for an existing lease acquisition to exist. We should just go ahead and acquire the lease (even if a recent acquisition happened) there is no harm in making another acquisition. You can get rid of the willExpire() method

Reply:
Yea, I just realized because this will definitely be running before a routine synchronously tries to acquire a new lease as part of `acquire` so the only situation where this would renew immediately after would be if `acquireFreshestFromStoreLocked` executes exactly when the timer fires.

I'll remove this.